Access entries in a FASTA file by index (nr of entry in the file)

HenningTimm opened the pull request:
I came across the need to access a FASTA file using a .fai index by position instead of by entry name. Example:

    let mut seq = Vec::new();
    reader.read_all_by_index(42, &mut seq).unwrap();

instead of

    let mut seq = Vec::new();
    reader.read_all("some fasta entry", &mut seq).unwrap();

When reading in the .fai index, the names are already saved in the `seqs` vector of the index. The index can be used to get the name from the `seqs` vector and then proceed as before (i.e. get an index object using `Index.idx(name)`). I implemented a `_by_index` version for `read`, `read_all`, `read_iter`, and `read_all_iter` as well as tests.

I am not happy with the names of the methods though. The `by_index` part could be perceived to refer to the .fai index itself instead of the position of the entry in the file. However, `by_position`, for me, would sound more like a (byte) position in the file. I am (very) open for suggestions here.

johanneskoester commented:
Thanks! I am a bit unsure about the naming as well. I will think about it...

johanneskoester commented:
Would it make sense to have a method `fn fetch_by_rid(&mut self, rid: usize)`, and `fn fetch(&mut self, name: &str)` that sets the selected sequence for the reader? And then, all other methods are the same for both cases?

HenningTimm commented:
That would make accessing the sequence a two step process, right? Something like this:

    let mut seq = vec![];
    reader.fetch(&"chr42");
    reader.read_all(&mut seq);

    reader.fetch_by_rid(23);
    let seq_it = reader.read_iter(17, 56);

**Pro:** 
* Few functions (`fetch`, `fetch_by_rid`, `read`, `read_all`, `read_iter`, `read_all_iter`)
* Nicer function names

**Con:**
* One additional line of code
* Usability might not be obvious (The missing return type should make that clear though.)
* (API change)


## Another Option ##
I am not familiar with idiomatic Rust, so I am not sure if this is a good Idea, but another option could be to introduce an enum:

    enum SeqAddress <'a>{
        Name(&'a str),
        Rid(usize),
    }

and replace `seqname: &str` with `seq_address: SeqAddress`:

    let mut seq = vec![];
    reader.read_all(SeqAddress::Name("chr42"), seq);

Would this be preferable?

johanneskoester commented:
I must admit that I am unsure... The enum is also an interesting solution. The fetch is what people are used to from htslib...

johanneskoester commented:
In a way, the fetch is also idiomatic in rust, see [here](https://doc.rust-lang.org/beta/std/io/trait.Seek.html). I think I prefer that. However, we might want to also put the interval into the fetch method.

HenningTimm commented:
I updated the code and added two new tests that check if reading fails if no sequences has been fetched.

I am not sure if `start`, `stop` and `fetched_idx` should be reset after a successful read. I will think about that.

johanneskoester commented on src/io/fasta.rs:
```diff
@@ -321,7 +316,7 @@ impl<R: io::Read + io::Seek> IndexedReader<R> {
-    fn idx_by_index(&self, seq_index: usize) -> io::Result<IndexRecord> {
+    fn idx_by_rid(&self, seq_index: usize) -> io::Result<IndexRecord> {
         match self.index.seqs.get(seq_index) {
             Some(name) => self.idx(&name),
```
Mhm, this looks a bit weird to me. We provide an index, then we get a name, and with the name we look up an index again? Why that?

HenningTimm replied:
The main data structures in the `Index` struct used by `IndexedReader` are a HashMap that maps sequence names to idx objects and a Vector of sequence names.

    pub struct Index {
        inner: collections::HashMap<String, IndexRecord>,
        seqs: Vec<String>,
    }

I chose this to minimize the additional memory required and to minimize changes to the code. Alternatively we could add a `Vec<IndexRecord>` or `HashMap<usize, IndexRecord>` to directly map positions to index record. This would eliminate one layer of indirection at the expense of more memory.

However, this might be a naming issue. I renamed seq_index to `rid`.

johanneskoester replied:
What about `inner: Vec<IndexRecord>` and `name_to_rid: HashMap<String, usize>`. This gives us constant time access for both RID and name.

HenningTimm replied:
Done. This made the `sequences` method a little more complex (now sorting is required to restore the order of the records in the file).

Querying the Index with an RID is just indexing into the `inner` Vector, while querying with a name requires a HashMap lookup to get the rid and then index into the Vector.

johanneskoester replied:
Can't you add the name to the index record struct, and let the vector be a vector of references to the index records from the hashmap?

johanneskoester commented on src/io/fasta.rs:
```diff
@@ -96,29 +97,30 @@ impl<R: io::Read> Reader<R> {
-    inner: collections::HashMap<String, IndexRecord>,
-    seqs: Vec<String>,
+    inner: Vec<IndexRecord>,
+    name_to_rid: collections::HashMap<String, usize>,
```
This looks perfect. Now, if you add the name additionally to the IndexRecord, the sequences method becomes very easy, without sorting.

johanneskoester commented:
Very nice work! Please also add yourself to the list of contributors and put your name in the license notice at the top. Thank you!
